8253694: Remove Thread::muxAcquire() from ThreadCrashProtection()

pchilano opened the pull request:
Hi all,

Please review the following patch. Current ThreadCrashProtection() implementation uses static members which requires the use of Thread::muxAcquire() to allow only one user at a time. We can avoid this synchronization requirement if each thread has its own ThreadCrashProtection *data.
I tested it builds on Linux, macOS and Windows. Since the JfrThreadSampler is the only one using this I run all the tests from test/jdk/jdk/jfr/. I also run some tests with JFR enabled while forcing a crash in OSThreadSampler::protected_task() and tests passed with several "Thread method sampler crashed" UL output. Also run tiers1-3.

Thanks,
Patricio

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8253694](https://bugs.openjdk.java.net/browse/JDK-8253694): Remove Thread::muxAcquire() from ThreadCrashProtection()


### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Daniel D. Daugherty](https://openjdk.java.net/census#dcubed) (@dcubed-ojdk - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to c3168cc7d936e2b23f9fc6eb7fe6566fec4b054d


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/376/head:pull/376`
`$ git checkout pull/376`

pchilano commented:
/label add hotspot-jfr

coleenp commented on src/hotspot/share/runtime/thread.hpp:
```diff
@@ -737,6 +737,23 @@ class Thread: public ThreadShadow {
+    }
+  }
+#endif
+
```
Could this be pushed down into osThread ?

pchilano replied:
Yes, it can but it's not that clean I think given the osthread indirections. Here's how it looks: https://github.com/pchilano/jdk/commit/73a41ad867b4b7466cdddc87173828b4e80f8179

I think another alternative could be to remove the "#ifndef _WINDOWS" clause and define an empty check_crash_protection() method in os_windows.hpp

dholmes-ora reviewed (commented):
Hi Patricio,

This seems fine, but I'm wondering what the motivation for this change was? Adding more per-thread state is arguably just adding to the clutter of per-thread state. I don't know if this approach was considered when @robehn fixed JDK-8183925.

Thanks,
David

robehn commented:
I don't think so.

I have not seen crash protection catching anything since pre-JDK 9.
I would remove it completely instead :) Not sure what JFR team says...

mgronlun commented:
Like David I would also like to know more about the motivation. Is the feature expected to be used by a larger number of threads? If so, there might be concerns about scalability that was not considered initially.

I agree that we have seen less, and for a long time almost no, asserts related to thread sampling in our testing with fastdebug builds (only product builds run with the protection). At the same time, I am not sure how representative that is considering all the code that is out there.  We should also keep in mind that we have upcoming features that will have slightly different stack layouts which will affect how stackwalking is achieved, so I would recommend keeping the established safety mechanism.

pchilano commented:
I looked at the users of Thread::muxAcquire/muxRelease and this was one of the two places where it is used. If we are going to have a crash protection mechanism for general use then the fields should not be static. Now, if we know only the JfrThreadSampler uses it and we want to optimize away that pointer in the thread object then that makes sense, but then we should remove _crash_mux.

pchilano commented:
Please see the update. Since there was concern about adding a new pointer in the Thread class I kept the fields as static and just removed _crash_mux. I also fixed the comment about the class being used by the JfrSampler instead of the Watcher thread.

dcubed-ojdk commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -4999,9 +4999,10 @@ void os::pause() {
-volatile intptr_t os::ThreadCrashProtection::_crash_mux = 0;
 
 os::ThreadCrashProtection::ThreadCrashProtection() {
+  assert(Thread::current()->is_JfrSampler_thread(), "should be JFRSampler");
+  _protected_thread = Thread::current();
```
Perhaps this:
  `_protected_thread = Thread::current();`
  `assert(_protected_thread->is_JfrSampler_thread(), "should be JFRSampler");`
would be a little more clean.

dcubed-ojdk reviewed (approved):
Looks good. I only have minor suggested changes, but it's you call
on whether to make those changes.

pchilano commented:
> Looks good. I only have minor suggested changes, but it's your call
> on whether to make those changes.

Thanks Dan! Updated.

dcubed-ojdk reviewed (approved):
Thumbs up.

coleenp reviewed (approved):
LGTM!

dholmes-ora reviewed (approved):
So let me see if I've got this straight. Prior to JDK-8183925 CrashProtection was exclusively for the WatcherThread. JDK-8183925 generalized that to allow any(?) thread to use it. Now as the only client is the JfrSampler thread we are making crash protection exclusively only available to it.

pchilano commented:
> So let me see if I've got this straight. Prior to JDK-8183925 CrashProtection was exclusively for the WatcherThread. JDK-8183925 generalized that to allow any(?) thread to use it. Now as the only client is the JfrSampler thread we are making crash protection exclusively only available to it.

Exactly.

pchilano commented:
Thanks @dcubed-ojdk, @coleenp and @dholmes-ora for the reviews.

pchilano commented:
/integrate
